Repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
Commit ID: 0f3f1610b15117a0e79e7fafe93cf0302e32c0fc

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import *
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import *

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
slime_data = pd.read_csv("science_data_large.csv")
print("First 15 rows of data:")
print(slime_data.head(15))
print("Summary:")
print(slime_data.info())

First 15 rows of data:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
Summary:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
features = slime_data[["Temperature °C", "Mols KCL"]]
label = slime_data["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9, random_state=42)

## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame([[545, 637]], columns=X_train.columns)
predicted = reg.predict(sample)
print("Predicted Output:")
print(predicted)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = reg.score(X_test, y_test)
print("\nScore:")
print(score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print("\nCoefficients:")
print(reg.coef_)
print("\nIntercept")
print(reg.intercept_)

Predicted Output:
[720485.07305574]

Score:
0.8552472077276095

Coefficients:
[ 866.14641337 1032.69506649]

Intercept
-409391.47958340764


The score gives a numeric value of how well the model captures a significant portion of the relationship between the inputs and the output. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of y, disregarding the input features, would get a $R^2$ score of 0.0. Therefore, since the score for the model was calculated to be about 0.8552, it means the model represents 85.52% of the target data based on the input features and fits the data relatively well.

Equation for hypothesis function:

$h(Temperature °C, Mols KCL) = 866.15\cdot Temperature °C + 1032.70\cdot Mols KCL - 409391.48$

## Part 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cv_score = cross_val_score(reg, features, label, cv=10)
print("\nCross-validation scores:")
print(cv_score)

# Report on their finding and their significance
print("\nMean:")
print(cv_score.mean())
print("\nStandard deviation:")
print(cv_score.std())


Cross-validation scores:
[0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]

Mean:
0.8552450341984701

Standard deviation:
0.031528762965342454


The cross-validation scores's mean of about 0.8552 is practically the same and corroborates the original score found by the score function in the above section. The minimum score was about 0.784 and the maximum score was about 0.887 which is a difference of about 0.1. As mentioned above, the closer the score is to 1.0, the better the model and these scores verify that the model performs well for the set of data since these scores are fairly high. 

## Part 5. Using Polynomial Regression

In [15]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(features)
X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(X_poly, label, test_size=0.1, train_size=0.9, random_state=42)
poly_model = LinearRegression().fit(X_train_poly, y_train_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.
sample_poly = poly.transform(sample)
predicted_poly = poly_model.predict(sample_poly)
print("\nPredicted Output:")
print(predicted_poly)

score_poly = poly_model.score(X_test_poly, y_test_poly)
print("\nScore:")
print(score_poly)

print("\nPolynomial Coefficients:")
print(poly.get_feature_names_out())
print(poly_model.coef_)


print("\nPolynomial Intercept:")
print(poly_model.intercept_)


Predicted Output:
[712463.39999271]

Score:
1.0

Polynomial Coefficients:
['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']
[ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]

Polynomial Intercept:
2.0479375962167978e-05


The score of 1.0 is the highest score you can get for a model. This means the model performs perfectly for the dataset and predicts target data with 100% accuracy based on the input features.

Equation for hypothesis function: 

$h(Temperature °C, Mols KCL) = 12\cdot Temperature °C + 2\cdot Temperature °C \cdot Mols KCL + 2.86\times 10^{-2}\cdot (Mols KCL)^2 + 2.05\times 10^{-5}$